### PR TITLE
Fix system cannot find file errors with subprocess.Popen

### LINKED ISSUE #49: Windows 10 Python 3.[WinError 2] The system cannot find the file specified
**OS:** Windows 10
**python version:** 3.9.11
**glpi-project/tools tag:** 0.4.5
**plugin-release version:** 1.0.5 (not helpful)

_Note: plugin-release script version has not changed since `glpi-project/tools` tag 0.1.7. The plugin-release script version is still 1.0.5, despite numerous modifications to fix issues with python version compatibility_.

All occurrences of `subprocess.Popen` in `tools/plugin-release` require additional argument `shell=True` otherwise we get the following error below on Windows with python 3.9.x.

```
Adding vendor libraries
Traceback (most recent call last):
  File "C:\Users\105061136\Documents\GitHub\ldapcomputers\vendor\glpi-project\tools\tools\plugin-release", line 754, in <module>
    main()
  File "C:\Users\105061136\Documents\GitHub\ldapcomputers\vendor\glpi-project\tools\tools\plugin-release", line 751, in main
    _do_build(repo, buildver)
  File "C:\Users\105061136\Documents\GitHub\ldapcomputers\vendor\glpi-project\tools\tools\plugin-release", line 320, in _do_build
    prepare(plugin_name, archive)
  File "C:\Users\105061136\Documents\GitHub\ldapcomputers\vendor\glpi-project\tools\tools\plugin-release", line 417, in prepare
    p1 = subprocess.Popen(
  File "C:\Users\105061136\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\105061136\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 1420, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

### PR DESCRIPTION
**Pull request by derry43**
All subprocess.Popen calls get shell-True argument. Fixes #49.

### CODE REVIEW

**Review by trasher — DISMISSED** [review]
Seems OK to me

**Review by cedric-anne — CHANGES_REQUESTED** [review]
It breaks the script on Alpine linux.

Looks like command arguments are no longer used.

image
image

**Comment by derry43** [comment]
> subprocess.Popen

Hmm.... don't have a unix system but just found a note on subprocess documentation which describes some platform differences in the manner in which `shell=True` affects argument processing on Unix systems. This also suggests that their is no need to specify `shell=True` on windows unless the command is a shell built-in command, despite the evidence to the contrary. Going to dig into this further.

> On Windows, if _args_ is a sequence, it will be converted to a string in a manner described in [Converting an argument sequence to a string on Windows](https://docs.python.org/2/library/subprocess.html#converting-argument-sequence). This is because the underlying `CreateProcess()` operates on strings.
> 
> The shell argument (which defaults to `False`) specifies whether to use the shell as the program to execute. If shell is `True`, it is recommended to pass args as a string rather than as a sequence.
> 
> On Unix with `shell=True`, the shell defaults to `/bin/sh` If args is a string, the string specifies the command to execute through the shell. This means that the string must be formatted exactly as it would be when typed at the shell prompt. This includes, for example, quoting or backslash escaping filenames with spaces in them. If _args_ is a sequence, the first item specifies the command string, and **_any additional items will be treated as additional arguments to the shell itself_**. That is to say, [Popen](https://docs.python.org/2/library/subprocess.html#subprocess.Popen) does the equivalent of:
> 
> ```Popen(['/bin/sh', '-c', args[0], args[1], ...])```
> 
> On Windows with `shell=True`, the `COMSPEC` environment variable specifies the default shell. The only time you need to specify `shell=True` on Windows is when the command you wish to execute is built into the shell (e.g. **dir** or **copy**). You do not need `shell=True` to run a batch file or console-based executable.

**Comment by derry43** [comment]
Composer on Windows is a batch file and it looks like the issue that when `subprocess.Popen`  attempts to run a command `composer` it will only try to run an executable called `composer.exe` or `composer.com` but will not launch a batch file called `composer.bat` without using a shell.

**Review by trasher — APPROVED** [review]

File: tools/plugin-release [hunk 1]
```diff
@@ -409,11 +420,19 @@ def prepare(rel_name, archive):
...
     if os.path.exists(os.path.join(build_dir, 'composer.json')):
-        composer = ['composer', 'install', '-o', '--no-dev']
+
+        composer_cmd = 'composer'
+        if os.name=='nt':
+            if not cmd_exists([composer_cmd, '--version']):
+                composer_cmd = 'composer.bat'
+        if not cmd_exists([composer_cmd, '--version']):
+            print_err('composer not found!')
+            sys.exit(1)
```
Comment (cedric-anne) [inline tools/plugin-release:430]: I do not really know how `os.name` is computed, but I do not like relying on such stuffs (maybe due to unstabillity of such information on Javascript env).

```python
composer_cmd = 'composer'
if not cmd_exists([composer_cmd, '--version']):
    composer_cmd = 'composer.bat'
    if not cmd_exists([composer_cmd, '--version']):
        print_err('composer not found!')
        sys.exit(1)
```

Comment (trasher) [reply tools/plugin-release:430]: Relying on OS name is indeed not required

**Comment by cedric-anne** [comment]
I cherry-picked, squashed and adapted your commits in 540fdc0391135c8e45d8f1caf48f36c59b89a84b .

Thanks for the contribution.

**Pull request closed without merging**